lib/galaxy/tools/actions/__init__.py

```diff
@@ -748,17 +751,24 @@ class ObjectStorePopulator(object):
     def __init__(self, app):
         self.object_store = app.object_store
         self.object_store_id = None
+        self.objects = []
 
-    def set_object_store_id(self, data):
+    def register(self, data):
         # Create an empty file immediately.  The first dataset will be
         # created in the "default" store, all others will be created in
         # the same store as the first.
-        data.dataset.object_store_id = self.object_store_id
+        dataset = data.dataset
+        dataset.object_store_id = self.object_store_id
+        self.objects.append(dataset)
         try:
-            self.object_store.create(data.dataset)
+            self.object_store.set_object_store_id(dataset)
         except ObjectInvalid:
-            raise Exception('Unable to create output dataset: object store is full')
-        self.object_store_id = data.dataset.object_store_id  # these will be the same thing after the first output
+            raise Exception('Unable to set object store id for output dataset: object store is full')
+        self.object_store_id = dataset.object_store_id  # these will be the same thing after the first output
+
+    def create_datasets(self):
+        for obj in self.objects:
+            self.object_store.create(obj)
```

Comment:
I think the `try:` / `except ObjectInvalid:` should be moved from the `register()` method above to around this line.

Reply:
But isn't the most likely and first place it gets thrown from set_object_store_id?

```python
    def set_object_store_id(self, obj, **kwargs):
        if obj.object_store_id is None:
            try:
                obj.object_store_id = random.choice(self.weighted_backend_ids)
            except IndexError:
                raise ObjectInvalid('objectstore.create, could not generate '
                                    'obj.object_store_id: %s, kwargs: %s'
                                    % (str(obj), str(kwargs)))
```

create could thrown ObjectInvalid - but it will for other reasons not related to full disks based on my reading:

```
        if extra_dir and extra_dir != os.path.normpath(extra_dir):
            log.warning('extra_dir is not normalized: %s', extra_dir)
            raise ObjectInvalid("The requested object is invalid")
        # ensure that any parent directory references in alt_name would not
        # result in a path not contained in the directory path constructed here
        if alt_name and not safe_relpath(alt_name):
            log.warning('alt_name would locate path outside dir: %s', alt_name)
            raise ObjectInvalid("The requested object is invalid")
```

The ``set_object_store_id`` method seems like it would throw it if ``weighted_backend_ids`` becomes empty which it could based on file system monitoring so it seems like the right place to catch and throw that message?

Reply:
That looks correct, thanks for checking!